### PR TITLE
feat(icon): Add support to svg symbol icon set

### PR DESCRIPTION
**Pull request by cobisimo**
## PR Checklist
Please check that your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently icon sets with symbols are not supported. Icon don't display as symbols are not rendered.

Issue Number: #8689

## What is the new behavior?

This is kind of backport from material2 (angular 2 material design) project. You can check this:
https://github.com/angular/material2/blob/54252d5c1be882e6ffaec0f72516363f66b9de58/src/lib/icon/icon-registry.ts#L410
If detect symbol element iinstead svg within icon set convert it to svg.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by Splaktar** [comment]
Thank you for your contribution. Please squash your commits and update the commit message to follow the [commit message guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format). This is important for your changes to properly show up in the [changelog](https://github.com/angular/material/blob/master/CHANGELOG.md).

**Comment by Splaktar** [comment]
The description of the current behavior here and the related fix in this PR don't seem to map directly to https://github.com/angular/material/issues/8689. It does seem related, but it doesn't seem to be the same exact issue.

Can you please provide a CodePen that demonstrates the issue being solved here? The code seems to make sense, but it's not clear that there is an issue without a demo that reproduces the issue.

**Comment by cobisimo** [comment]
This PR is not direct fix for #8689 issue, but related to it. It is directly related to this closed issue #1514 (I didn't wont to do PR against closed ticket, but can change if you think it is better).
The issue here is that md-svg-icon use unique svg sprite `<svg>` packed within `<svg>`. There are multiple svg icon sprite generators like svg-sprite-loader, webpack-svgstore-plugin, gulp-svg-sprites ... They all pack icons as `<symbol>` within `<svg>`. If we try to use such icon pack it will not work, as `<symbol>` get embeded in `<md-icon>` instead `<svg>` and symbols are not renderable. This fix make addition to present setup to detect if source icon map use symbols and transfer them to svgs on the fly while creating icon cache.
Usual use of symbols to render icons is using `<use xlink:href="#icon-name"></use>`. This approach is good as it don't require icon cache, but switch to this require more changes to md-icon and have issue working on some browsers. As I said I didn't apply this logical way, I make a mix, using symbol within present setup, the similar way as it is done in material2.

**Comment by Splaktar** [comment]
OK, thank you for the link to #1514. That's good to know.

We'll certainly need a unit test to cover this behavior. 

It might be worth looking into adding a demo for this in the icon demos as well. If the demo gets too complicated or you have issues with it, please let me know.

**Comment by cobisimo** [comment]
@Splaktar I added unit tests and demo with 2 additional icons to `Demos -> Icons -> Svg Icon Sets` block. Also commits should be squashed.

**Comment by Splaktar** [comment]
@cobisimo thanks for adding the tests and demos. I'll take a look.

It looks like the commits didn't get squashed properly. You'll need to force push your branch to GitHub after squashing in order for things to work properly.

File: src/components/icon/js/iconService.js [hunk 1]
```diff
@@ -575,7 +575,14 @@ function MdIconService(config, $templateRequest, $q, $log, $mdUtil, $sce) {
...
   function Icon(el, config) {
-    if (el && el.tagName != 'svg') {
+    // If the node is a <symbol>, it won't be rendered so we have to convert it into <svg>.
+    if (el && el.tagName.toLowerCase() === 'symbol') {
+      var viewbox = el.getAttribute('viewBox');
+      el = angular.element('<svg xmlns="http://www.w3.org/2000/svg">').html(el.innerHTML)[0];
+      if (viewbox) el.setAttribute('viewBox', viewbox);
+    }
+
+    if (el && el.tagName.toLowerCase() != 'svg') {
```
Comment (Splaktar) [inline src/components/icon/js/iconService.js]: Please use `!==` here.

File: docs/app/img/icons/sets/symbol-icons.svg [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<svg>
+<symbol id="spinner" viewBox="0 0 44 44"><path d="M5.386 19.817c-.584 4.44.595 8.84 3.32 12.39 5.628 7.333 16.172 8.722 23.504 3.095 7.332-5.628 8.72-16.173 3.093-23.504-3.57-4.65-9.406-7.12-15.238-6.447-.264.028-.51-.16-.54-.428-.03-.267.16-.51.43-.54 6.166-.707 12.345 1.902 16.12 6.822 5.956 7.76 4.488 18.917-3.27 24.87-3.215 2.467-7.01 3.66-10.78 3.66-5.33 0-10.604-2.385-14.092-6.932C5.048 29.042 3.8 24.386 4.42 19.69c.595-4.532 2.86-8.578 6.392-11.447H5.57c-.27 0-.488-.22-.488-.488 0-.27.218-.488.487-.488h6.678c.27 0 .488.218.488.488v6.68c0 .27-.218.487-.488.487s-.488-.217-.488-.488V8.737c-3.53 2.724-5.792 6.657-6.374 11.08z"><animateTransform attributeName="transform" type="rotate" from="0 22 22" to="360 22 22" dur="3s" repeatCount="indefinite"/></path></symbol>
```
Comment (Splaktar) [inline docs/app/img/icons/sets/symbol-icons.svg]: I like the AngularJS logo SVG! Where does this spinner come from? Did you create it or do you have a link to the related license?

**Comment by Splaktar** [comment]
Oh, one more thing, in your squashed commit, please use `feat(icon):` instead of `feature(icon):` so that our changelog will pick it up correctly.

File: docs/app/img/icons/sets/symbol-icons.svg (hunk 2 above)
Comment (cobisimo) [reply docs/app/img/icons/sets/symbol-icons.svg]: I was using icon from webpack-svgstore-plugin. I change it now to material sync icon: https://material.io/icons/#ic_sync to not guess about licence.

Comment (Splaktar) [reply docs/app/img/icons/sets/symbol-icons.svg]: Excellent, thank you.

**Pull request merged**